File: Data Structures & Algorithms/count-paths/submission-2.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<int> prev(n, 0);
        for (int r = m - 1; r >= 0; --r) {
            vector<int> curr(n, 0);
            curr[n - 1] = 1;
            for (int c = n - 2; c >= 0; --c) {
                curr[c] = curr[c + 1] + prev[c];
            }
            prev = curr;
        }
        return prev[0];
    }
};

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        prev_group = dummy
        while True:
            k_node = prev_group
            for _ in range(k):
                k_node = k_node.next
                if k_node is None:
                    return dummy.next
            next_group = k_node.next
            prev = k_node.next
            curr = prev_group.next
            while curr != next_group:
                curr_next = curr.next
                curr.next = prev
                prev = curr
                curr = curr_next
            tmp = prev_group.next
            prev_group.next = prev
            prev_group = tmp
        return dummy.next